### PR TITLE
feat: init support mvn -D verbose support

### PR DESCRIPTION
**Pull request by anthogez**
Co-authored-by: Or Sagie or@snyk.io

With this pull request if an user asks for -Dverbose arg that invalidates -DoutputType=dot
We will still digest the mvn cli output despite no longer having a digraph DOT format but a mvn tree.
We generate depGraph/s for single and/or aggregate projects.

### CODE REVIEW

File: lib/depgraph-from-parsetree.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+export function buildDepGraphFromParsedTree(parsedTree: ParsedNode): DepGraph {
+    const builder = new DepGraphBuilder(
+        { name: 'maven' },
+        {
+            name: `${parsedTree.dependency.groupId}:${parsedTree.dependency.artifactId}`,
+            version: parsedTree.dependency.version,
+        },
+    );
+
+    const addDependenciesRecursively = (node: ParsedNode, parentId: string) => {
```
Comment (anthogez) [inline lib/depgraph-from-parsetree.ts:13]: todo fix this ugly thing

File: lib/parse/stdout/index.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+export function parseStdout(stdout: string, verbose = false): string[] {
+  if (errorLabel.test(stdout) && !successLabel.test(stdout)) {
+    throw new Error('Maven output contains errors.');
+  }
+  const withoutLabels = stdout.replace(logLabel, '');
+  const parsedOutputs = verbose
+    ? parseTree(withoutLabels)
+    : parseDigraph(withoutLabels);
+  if (!parsedOutputs) {
+    throw new Error('Cannot find any digraphs.');
```
Comment (37IulianPopovici) [inline lib/parse/stdout/index.ts:18]:
```suggestion
    throw new Error(verbose ? 'Cannot find any trees.' : 'Cannot find any digraphs.');
```

File: lib/index.ts [hunk 3]
```diff
@@ -179,7 +179,10 @@ export async function inspect(
         cwd: mvnWorkingDirectory,
       },
     );
-    const parseResult = parse(result, options.dev);
+
+    const includesVerboseFlag = mvnArgs.includes('-Dverbose');
```
Comment (gitphill) [inline lib/index.ts:183]: not very likely, but what if the command is `-Dverbose=false`, which is a user trying to disable verbose

File: lib/parse/tree.ts
Comment (gitphill) [inline lib/parse/tree.ts:1]: I wasn't expecting to see us reimplement so much. Can we not use the existing maven-graph-builder, and adapt dep-graph to full traverse all lines (don't prune) when verbose is being used?

File: lib/parse/stdout/index.ts (hunk 2 above)
Comment (anthogez) [reply lib/parse/stdout/index.ts]: Thanks, done

File: tests/jestified/mvn-parsed-tree-to-depgraph.spec.ts [hunk 4]
```diff
@@ -0,0 +1,191 @@
+import { parseTrees } from '../../lib/parse/tree';
+import { createFromJSON } from '@snyk/dep-graph';
+import { parseStdout } from '../../lib/parse/stdout';
+import {
+  smallVerboseDepGraphData,
+  largeVerboseDepGraphData,
+} from './fixtures/verbose-depgraph';
+import { verboseAggregatedProjects } from './fixtures/verbose-aggregated-projects';
+
+it('small - output from mvn dependency:tree -Dverbose to be parsed from tree to depgraph', () => {
```
Comment (gitphill) [inline tests/jestified/mvn-parsed-tree-to-depgraph.spec.ts:10]: this isn't testing reality is it? we use `-DoutputType=dot` which has a very different output

File: lib/parse/tree.ts
Comment (gitphill) [inline lib/parse/tree.ts:1]: I wasn't expecting us to reimplement so much. Can we not use `-Dverbose` with `-DoutputType=dot`?

File: tests/jestified/mvn-parsed-tree-to-depgraph.spec.ts (hunk 4 above)
Comment (gitphill) [reply tests/jestified/mvn-parsed-tree-to-depgraph.spec.ts:10]: (wheres the inspect test that would have uncovered this)

File: lib/parse/tree.ts
Comment (anthogez) [reply lib/parse/tree.ts:1]: Not in all versions of the plugin, for this reason we have introduced this change, even though our preference is always -DoutputType=dot

File: lib/parse/tree.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+      return parseDependency(omittedParts[0], omittedParts[1]);
+    }
+    if (line.indexOf('omitted for duplicate') > -1) {
+      const omittedParts = line.split(' - omitted for duplicate');
+      return parseDependency(omittedParts[0]);
+    }
+    return parseDependency(line);
+  }
+
+  private computeDepth(line: string): number {
```
[the remark this answers is not in the document]
Comment (anthogez) [reply lib/parse/tree.ts:146]: we have overly stressed this, as I said it's an incremental change for cases where mvn plugin does not support -Doutputype=dot with -DVerbose

File: lib/index.ts (hunk 3 above)
Comment (anthogez) [reply lib/index.ts:183]: this is a good one, we will make sure of double check that -Dverbose will be false 👍

File: lib/index.ts [hunk 6]
```diff
@@ -165,30 +170,32 @@ export async function inspect(
...
+
     return {
       plugin: {
         name: 'bundled:maven',
         runtime: 'unknown',
         meta: {
           versionBuildInfo: {
             metaBuildVersion: {
-              mavenVersion,
+              mavenFullVersion,
```
Comment (orsagie) [inline lib/index.ts:198]: change to mavenVersion: mavenFullVersion

**Comment by anthogez** [comment]
work continued here
https://github.com/snyk/snyk-mvn-plugin/pull/168
https://github.com/snyk/snyk-mvn-plugin/pull/169

**Pull request closed without merging**